### PR TITLE
Cache parameters

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/SciML/DataInterpolations.jl/issues/271 (the other half)

## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
That was a bit fiddly to get right (or at least pass the tests). Quick benchmark:

```julia
using DataInterpolations
using BenchmarkTools

N = 1000
u = rand(N)
t = cumsum(rand(N))

itp = LinearInterpolation(u, t)

t_eval = range(first(t), last(t), length = 10000)

@btime itp($t_eval)

# old: 162.200 μs (3 allocations: 78.23 KiB)
# new: 119.900 μs (3 allocations: 78.23 KiB)
```

**Comment by SouthEndMusic** [comment]
Another benchmark:

```julia
using DataInterpolations
using BenchmarkTools

N = 1000
u = rand(N)
t = cumsum(rand(N))

itp = QuadraticInterpolation(u, t)

@btime DataInterpolations.integral(itp, t[end])

# old: 18.400 μs (3 allocations: 48 bytes)
# new:  9.000 μs (3 allocations: 48 bytes)
```

**Comment by SouthEndMusic** [comment]
I'll leave it at caching parameters for `LinearInterpolation` and `QuadraticInterpolation` for this PR. 

An interesting one to cache parameters for next are `BsplineInterpolation` and `BsplineApprox`, as they now allocate at every  call in `spline_coefficients`:

```julia
using DataInterpolations
using BenchmarkTools

N = 1000
u = rand(N)
t = cumsum(rand(N))

t = [0, 62.25, 109.66, 162.66, 205.8, 252.3]
u = [14.7, 11.51, 10.41, 14.95, 12.24, 11.22]
A1 = QuadraticInterpolation(u,t)
A2 = BSplineInterpolation(u, t, 2, :Uniform, :Uniform)
A3 = BSplineApprox(u,t, 2, 4, :Uniform, :Uniform)

ts = collect(range(first(t), last(t), length = 10000))

@btime A1.(ts)
# 91.100 μs (3 allocations: 78.25 KiB)

@btime A2.(ts)
# 394.500 μs (10003 allocations: 1.14 MiB)

@btime A3.(ts)
# 387.400 μs (10003 allocations: 1015.77 KiB)
```

**Comment by SouthEndMusic** [comment]
Refactor of `LagrangeInterpolation` to be non-allocating and using `idx_prev`:

```julia
using DataInterpolations
using BenchmarkTools

N = 10
t = cumsum(rand(N))
u = rand(N)
itp = LagrangeInterpolation(u, t, N-1)

ts = range(first(t), last(t), length = 10000)

@btime itp($ts)
# old: 1.806 ms (20003 allocations: 2.82 MiB)
# new: 465.300 μs (3 allocations: 78.23 KiB)
```

**Comment by SouthEndMusic** [comment]
Using knowledge of which spline coefficients are non-zero:

```julia
using DataInterpolations
using BenchmarkTools

t = cumsum(rand(100))
u = rand(100)
A = BSplineInterpolation(u, t, 2, :Uniform, :Uniform)

ts = collect(range(first(t), last(t), length = 10000))

@btime A($ts)
# old: 1.161 ms (2 allocations: 78.17 KiB)
# new: 663.900 μs (2 allocations: 78.17 KiB)

```

File: src/online.jl [hunk 1]
```diff
@@ -1,40 +1,63 @@
 import Base: append!, push!
 
-function push!(di::LinearInterpolation{U, T}, u::eltype(U), t::eltype(T)) where {U, T}
-    push!(di.u, u)
-    push!(di.t, t)
-    di
+function push!(A::LinearInterpolation{U, T}, u::eltype(U), t::eltype(T)) where {U, T}
```
Comment (sathvikbhagavan) [inline src/online.jl:3]: Shouldn't this be allowed only if `safetycopy` is false?

File: test/interpolation_tests.jl [hunk 2]
```diff
@@ -75,14 +75,6 @@ end
     @test A(3.0) == 2.0
     @test A(4.0) == 3.0
 
-    u = [0.0, NaN, 2.0, 3.0]
```
Comment (sathvikbhagavan) [inline test/interpolation_tests.jl:78]: Why was this removed?

**Review by sathvikbhagavan — COMMENTED** [review]
Looks good! Can you also add a few tests to ensure all cache parameters are set properly?

File: test/interpolation_tests.jl (hunk 2 above)
Comment (SouthEndMusic) [reply test/interpolation_tests.jl:78]: Unless I am missing something very subtle, this block of tests is identical to the previous one

File: src/online.jl (hunk 1 above)
Comment (SouthEndMusic) [reply src/online.jl:3]: I don't think I fully understand the reasoning here. If `safetycopy` is true, then the `u, t` vectors within the interpolation object can safely be assumed to be only used within the context of this package. Surely then it is safe to add data to it? Only if `u, t` are also used in code outside the `DataInterpolations` context do you run the risk of breaking that code by modifying `u, t`.

File: test/interpolation_tests.jl (hunk 2 above)
Comment (sathvikbhagavan) [reply test/interpolation_tests.jl:78]: Ah yes, didn't see that.

File: src/online.jl (hunk 1 above)
Comment (sathvikbhagavan) [reply src/online.jl:3]: I think I understand your point. It comes down what should be the semantics of `push!` for interpolation objects. Currently it mutates the original array. Maybe we should handle both cases? If it is true, it would push to the copy, if it is false it would mutate the original array. For both we have to compute the parameters again. 
@ChrisRackauckas, your thoughts?

File: docs/src/interface.md [hunk 3]
```diff
@@ -35,6 +35,23 @@ A2(300.0)
     
     The values computed beyond the range of the time points provided during interpolation will not be reliable, as these methods only perform well within the range and the first/last piece polynomial fit is extrapolated on either side which might not reflect the true nature of the data.
 
+The keyword `safetycopy=false` can be passed to make sure no copies of `u` and `t` are made when initializing the interpolation object.
```
Comment (sathvikbhagavan) [inline docs/src/interface.md:38]:
```suggestion
The keyword `safetycopy = false` can be passed to make sure no copies of `u` and `t` are made when initializing the interpolation object.
```

File: docs/src/interface.md [hunk 4]
```diff
@@ -35,6 +35,23 @@ A2(300.0)
...
+
+```@example interface
+A3 = QuadraticInterpolation(u, t; safetycopy = false)
+
+# Check for same memory
+u === A3.u.parent
+```
+
+Note that this does not prevent allocation in every interpolation constructor call, because parameter values are cached for some interpolation types.
```
Comment (sathvikbhagavan) [inline docs/src/interface.md:47]: Can you also add a line describing which interpolations support caching?

File: src/online.jl [hunk 5]
```diff
@@ -1,40 +1,69 @@
...
+    !A.safetycopy && throw(UnsafeAddDataError())
```
Comment (sathvikbhagavan) [inline src/online.jl:4]: Is this needed? Shouldn't we allow for both?

File: test/integral_tests.jl [hunk 6]
```diff
@@ -83,7 +83,7 @@ end
 
 @testset "LagrangeInterpolation" begin
     u = [1.0, 4.0, 9.0]
-    t = [1.0, 2.0, 3.0]
+    t = [1.0, 2.0, 6.0]
```
Comment (sathvikbhagavan) [inline test/integral_tests.jl:86]: Why was this changed?

File: test/online_tests.jl [hunk 7]
```diff
@@ -1,18 +1,31 @@
 using DataInterpolations
 
-t = [1, 2, 3]
-u = [0, 1, 0]
+t1 = Float64[1, 2, 3]
+u1 = Float64[0, 1, 0]
```
Comment (sathvikbhagavan) [inline test/online_tests.jl:4]:
```suggestion
t1 = [1.0, 2.0, 3.0]
u1 = [0.0, 1.0, 0.0]
```

File: test/online_tests.jl [hunk 8]
```diff
@@ -1,18 +1,31 @@
...
 
-for di in [LinearInterpolation, QuadraticInterpolation, ConstantInterpolation]
-    li = di(copy(u), copy(t))
-    append!(li, u, t)
-    li2 = di(vcat(u, u), vcat(t, t))
-    @test li.u == li2.u
-    @test li.t == li2.t
+t2 = Float64[4, 5, 6]
+u2 = Float64[1, 2, 1]
```
Comment (sathvikbhagavan) [inline test/online_tests.jl:7]:
```suggestion
t2 = [4.0, 5.0, 6.0]
u2 = [1.0, 2.0, 1.0]
```

File: test/integral_tests.jl (hunk 6 above)
Comment (SouthEndMusic) [reply test/integral_tests.jl:86]: I added this function to conveniently test whether the implementation of a new interpolation type is complete: https://github.com/SciML/DataInterpolations.jl/pull/274/files#diff-f66955aae2aee262495ff068999fe35f8501cd1cbef7b1057aa26d2f1099a4f3R6-R17

To do this, I moved throwing an error when integration is not defined from specialised methods of `integral` to methods of `_integral`. However, that now means that the extrapolation error is thrown before the 'integral not found' error. So, to still test the integral not found error, I had to make sure that the test for this does not try to extrapolate.

File: src/online.jl (hunk 5 above)
Comment (SouthEndMusic) [reply src/online.jl:4]: That is fine by me, I think it is clear enough that this leads to mutation of the input `u, t` if someone explicitly sets `safetycopy = false`.

File: src/interpolation_utils.jl [hunk 9]
```diff
@@ -1,14 +1,42 @@
...
+            A.idxs[i] = idx_max
+        end
+    end
+    return idx
 end
 
-function spline_coefficients(n, d, k, u::Number)
-    N = zeros(eltype(u), n)
+function spline_coefficients!(N, d, k, u::Number)
+    N .= 0
```
Comment (sathvikbhagavan) [inline src/interpolation_utils.jl:33]:
```suggestion
    N .= zero(u)
```

File: docs/src/interface.md [hunk 10]
```diff
@@ -35,6 +35,23 @@ A2(300.0)
...
+```@example interface
+A3 = QuadraticInterpolation(u, t; safetycopy = false)
+
+# Check for same memory
+u === A3.u.parent
+```
+
+Note that this does not prevent allocation in every interpolation constructor call, because parameter values are cached for all interpolation types except `ConstantInterpolation`.
+
+Because of the caching of parameters which depend on `u` and `t`, this data should not be mutated. Therefore `u` and `t` are wrapped in a `ReadOnlyArray`.
```
Comment (sathvikbhagavan) [inline docs/src/interface.md:49]:
```suggestion
Because of the caching of parameters which depend on `u` and `t`, this data should not be mutated. Therefore `u` and `t` are wrapped in a `ReadOnlyArray` from [ReadOnlyArrays.jl](https://github.com/JuliaArrays/ReadOnlyArrays.jl).
```

**Review by sathvikbhagavan — COMMENTED** [review]
Almost there. Just a few cleanups.

**Review by sathvikbhagavan — APPROVED** [review]
LGTM! Nice work! Thanks!

**Review by ChrisRackauckas — APPROVED** [review]
